### PR TITLE
User management e2e tests

### PR DESCRIPTION
**Pull request by sbryzak**
e2e tests for https://issues.redhat.com/browse/CRT-387

Signed-off-by: Shane Bryzak <sbryzak@gmail.com>

### CODE REVIEW

File: test/e2e/base_user_test.go [hunk 1]
```diff
@@ -0,0 +1,102 @@
...
+import (
+	"context"
+	"github.com/stretchr/testify/assert"
+
+	"github.com/codeready-toolchain/api/pkg/apis/toolchain/v1alpha1"
+	"github.com/codeready-toolchain/toolchain-e2e/testsupport"
+	"github.com/codeready-toolchain/toolchain-e2e/wait"
+	framework "github.com/operator-framework/operator-sdk/pkg/test"
+	"github.com/stretchr/testify/require"
+	"github.com/stretchr/testify/suite"
```
Comment (golangcibot) [inline test/e2e/base_user_test.go]: File is not `gofmt`-ed with `-s` (from `gofmt`)
```suggestion
	"github.com/stretchr/testify/suite"
	corev1 "k8s.io/api/core/v1"
```

**Comment by MatousJobanek** [comment]
I see that you're creating some new e2e tests but couldn't you just add two more subtests here: https://github.com/codeready-toolchain/toolchain-e2e/blob/master/test/e2e/e2e_test.go#L116
* one for the case when you set `UserSignup.Spec.Dectivated = true`
* and then the second one for the case when you reset back `UserSignup.Spec.Dectivated = false`

File: test/e2e/user_management_test.go [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+}
+
+func (s *userManagementIntegrationTest) TearDownTest() {
+	s.testCtx.Cleanup()
+}
+
+func (s *userManagementIntegrationTest) TestUserDeactivation() {
+	s.setApprovalPolicyConfig("automatic")
+
+	userSignup, mur := s.createAndCheckUserSignup(true, "iris-at-redhat-com", "iris@redhat.com", approvedByAdmin()...)
```
Comment (MatousJobanek) [inline test/e2e/user_management_test.go]: Please, don't create new user accounts (including all related resources) if it's not really needed - it makes the e2e tests even longer. Please check my comment https://github.com/codeready-toolchain/toolchain-e2e/pull/68#issuecomment-574756595 and consider adding it as a subtest to the main test located in `e2e_test.go`. 

I guess that it would be also good to verify the recreation of resources in case `UserSignup.Deactivated` is reset back to `false`

Comment (sbryzak) [reply test/e2e/user_management_test.go]: @MatousJobanek we are adding a number of user management-related features currently, see the "User Management Workflows" section at the end of the detailed design doc.  The nature of these features will necessitate the removal of all of a user's resources during testing in order to validate they are functioning correctly, and therefore the creation of new user accounts seems to be unavoidable.  Also since these features are quite comprehensive in nature it makes sense to separate their tests into a separate file - while we just have deactivation implemented so far there are a lot more tests coming shortly.

Comment (MatousJobanek) [reply test/e2e/user_management_test.go]: > we are adding a number of user management-related features currently, see the "User Management Workflows" section at the end of the detailed design doc.

I know about that and that's another reason why we should not unnecessarily create resources in the cases when it's not really needed.

> The nature of these features will necessitate the removal of all of a user's resources during testing in order to validate they are functioning correctly, and therefore the creation of new user accounts seems to be unavoidable.

But this can be easily done as part of the already existing test - we don't have to create a special test with all setup for it.

>  Also since these features are quite comprehensive in nature it makes sense to separate their tests into a separate file

These are e2e tests and we should verify the whole e2e flow - that means:
1. provision user account
2. let it working for some time
3. deactivate the user account
4. reactivate 

from my point of view, the best way to simulate the flow is by appending it to the main e2e test. 

> while we just have deactivation implemented so far there are a lot more tests coming shortly.

We all struggle with the e2e tests execution time. There are some Jira issues to improve it and some time ago we were also trying to minimize the number of tests where we provision the user accounts to a necessary minimum - we were merging some tests, deleting the duplicated ones and optimizing the flows. Adding new tests including the whole setup when we can add it to already existing ones is just a step back for me. 

I understand your reasoning - that having a separated test for the functionality looks logical, but this is not a unit test nor an integration test. Unfortunately, the e2e tests take significant time to be completed and we need to keep it as short as possible.

Just to be clear - all that I wrote above is only my point of view and my understanding, but if the team agrees on adding new tests like that, I'm fine with it ;-). WDYT @alexeykazakov @xcoulon @tinakurian

Comment (alexeykazakov) [reply test/e2e/user_management_test.go]: Initially I was for creating a separate big user management e2e test which would test all user management related flows. But @MatousJobanek's point of too long test execution is a valid point too.
Adding all the upcoming user tests (diactivaion, disabling, banning, etc) to the existing main e2e test could make it too big and hard to maintain.
On the other hand if creating a dedicated user management test significantly increases the execution time then it's probably not worth separating. But does it? I mean it will be slower for sure but how slower?

Comment (MatousJobanek) [reply test/e2e/user_management_test.go]: I'm fine with adding one big user management e2e test with several subtests, but not with adding new several e2e standalone tests for all the user management cases you mentioned. 

As for the times - I made some measuring:
* it takes around 15-17 seconds to set up the test suite 
* up to 2 seconds to create UserSignup & MUR - but I need to point out that in this case, we don't wait for all resources to be provisioned - in other words, we don't wait for `MUR.status == provisioned` (which we should IMO). If we wait for the complete provisioning, then it takes 5-6 seconds
* to tear down the test (delete all resources) it may take up to 4-5 seconds 

The biggest time takes to set up the test-suite. Knowing that we could technically improve the overall execution time by having only one test suite for all tests, but that is for another PR.
 
As for the particular tests - one additional test doesn't bring any significant delay - up to 10 seconds (for provisioning and cleanup). But if we added multiple new standalone tests, then the delay would linearly increase.
From the discussion and also from the code it looks like a that this PR creates a test suite base and in the future, we will add several standalone tests - that is something I'm trying to avoid.

Comment (alexeykazakov) [reply test/e2e/user_management_test.go]: > I'm fine with adding one big user management e2e test with several subtests, but not with adding new several e2e standalone tests for all the user management cases you mentioned.

Absolutely. I fully agree that we should not create standalone tests for each user management use case.
I was thinking about having two big test suites:
- the current e2e_test suite for the main flows (it's what we have now)
- a new user management suite with all relevant user management use cases covered.

Both suites are optimized and using as few stand-alone tests as possible. If possible it's just one big e2e flow in each test suite which covers all the use cases.

So, yes, +1 to what you said above @MatousJobanek.

Comment (MatousJobanek) [reply test/e2e/user_management_test.go]: Ok, so let's keep it separated, but please the next cases add as subtests to this user-management test.

File: test/e2e/base_user_test.go [hunk 3]
```diff
@@ -0,0 +1,102 @@
...
+	require.NoError(s.T(), err)
+
+	// Confirm the MUR was created
+	mur := s.assertCreatedMUR(userSignup)
+
+	return userSignup, mur
+}
+
+func (s *baseUserIntegrationTest) assertCreatedMUR(userSignup *v1alpha1.UserSignup) *v1alpha1.MasterUserRecord {
+	mur, err := s.hostAwait.WaitForMasterUserRecord(userSignup.Status.CompliantUsername)
```
Comment (MatousJobanek) [inline test/e2e/base_user_test.go]: Shouldn't we wait for the provisioning to be completed? To be sure that the user account is in a final state and all resources are created before deleting them?

File: test/e2e/user_management_test.go [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+
+	// Deactivate the user
+	userSignup.Spec.Deactivated = true
+	err := s.awaitility.Client.Update(context.TODO(), userSignup)
+	require.NoError(s.T(), err)
+	s.T().Logf("user signup '%s' set to deactivated", userSignup.Name)
+
+	err = s.hostAwait.WaitUntilMasterUserRecordDeleted(mur.Name)
```
Comment (MatousJobanek) [inline test/e2e/user_management_test.go]: It would be great if we could verify also the deletion of all other resources (UserAccount, User, Identity, Namespaces, etc....)

File: test/e2e/user_management_test.go [hunk 5]
```diff
@@ -0,0 +1,44 @@
...
+	require.NoError(s.T(), err)
```
Comment (MatousJobanek) [inline test/e2e/user_management_test.go]: Could you please add also a subtest when we reset `UserSignup.Deactivated` back to `false` and check that all resources are created again?

**Comment by MatousJobanek** [comment]
@sbryzak is there any plan to finish this PR?

File: test/e2e/base_user_test.go [hunk 6]
```diff
@@ -0,0 +1,118 @@
...
+	userSignup, err = s.hostAwait.WaitForUserSignup(userSignup.Name, wait.UntilUserSignupHasConditions(conditions...))
+	require.NoError(s.T(), err)
+
+	// Confirm the MUR was created
+	mur := s.assertCreatedMUR(userSignup)
+
+	return userSignup, mur
+}
+
+func (s *baseUserIntegrationTest) createAndCheckBannedUser(email string) (*v1alpha1.BannedUser) {
```
Comment (golangcibot) [inline test/e2e/base_user_test.go]: File is not `gofmt`-ed with `-s` (from `gofmt`)
```suggestion
func (s *baseUserIntegrationTest) createAndCheckBannedUser(email string) *v1alpha1.BannedUser {
```

File: test/e2e/user_management_test.go [hunk 7]
```diff
@@ -0,0 +1,147 @@
...
+
+	s.T().Run("user banning", func(t *testing.T) {
+		// when
+		s.setApprovalPolicyConfig("automatic")
+
+		// then
+		s.checkUserBanned()
+	})
+
+
```
Comment (golangcibot) [inline test/e2e/user_management_test.go]: File is not `gofmt`-ed with `-s` (from `gofmt`)
```suggestion

```

File: wait/host.go [hunk 8]
```diff
@@ -126,6 +129,35 @@ func (a *HostAwaitility) WaitForUserSignup(name string, criteria ...UserSignupWa
...
+func (a *HostAwaitility) WaitForBannedUser(email string) (bannedUser *toolchainv1alpha1.BannedUser, err error) {
+	md5hash := md5.New()
+	// Ignore the error, as this implementation cannot return one
+	_, _ = md5hash.Write([]byte(email))
+	emailHash := hex.EncodeToString(md5hash.Sum(nil))
+	labels := map[string]string{toolchainv1alpha1.BannedUserEmailHashLabelKey: emailHash}
+	opts := client.MatchingLabels(labels)
+
+	err = wait.Poll(a.RetryInterval, a.Timeout, func() (done bool, err error) {
+		obj := &toolchainv1alpha1.BannedUser{}
```
Comment (golangcibot) [inline wait/host.go]: ineffectual assignment to `obj` (from `ineffassign`)

File: test/e2e/user_management_test.go [hunk 9]
```diff
@@ -0,0 +1,147 @@
...
+	// Confirm the CompliantUsername has NOT been set
+	require.Empty(s.T(), userSignup.Status.CompliantUsername)
+
+	// Confirm that a MasterUserRecord wasn't created
+	_, err = s.hostAwait.WithRetryOptions(wait.TimeoutOption(time.Second * 10)).WaitForMasterUserRecord(username)
+	require.Error(s.T(), err)
+	return userSignup
+}
+
+func newBannedUser(t *testing.T, host *wait.HostAwaitility, email string) *v1alpha1.BannedUser {
```
Comment (golangcibot) [inline test/e2e/user_management_test.go]: `newBannedUser` - `t` is unused (from `unparam`)

File: test/e2e/user_management_test.go [hunk 10]
```diff
@@ -0,0 +1,147 @@
...
+		s.setApprovalPolicyConfig("automatic")
+
+		id := uuid.NewV4().String()
+		email := "testuser"+id+"@test.com"
+		s.createAndCheckBannedUser(email)
+
+		userSignup, _ := s.createAndCheckUserSignup(true, "testuser"+id, email, approvedAutomatically()...)
+
+		// Check the UserSignup is banned
+		userSignup, err := s.hostAwait.WaitForUserSignup(userSignup.Name, wait.UntilUserSignupHasConditions(approvedAutomaticallyButBanned()...))
```
Comment (golangcibot) [inline test/e2e/user_management_test.go]: SA4006: this value of `userSignup` is never used (from `staticcheck`)

File: test/e2e/user_management_test.go [hunk 11]
```diff
@@ -0,0 +1,147 @@
...
+		require.NoError(s.T(), err)
+	})
+}
+
+func (s *userManagementIntegrationTest) createUserSignupAndAssertAutoApproval(specApproved bool) (*v1alpha1.UserSignup, *v1alpha1.MasterUserRecord) {
+	id := uuid.NewV4().String()
+	return s.createAndCheckUserSignup(specApproved, "testuser"+id, "testuser"+id+"@test.com", approvedAutomatically()...)
+}
+
+func (s *userManagementIntegrationTest) createUserSignupAndAssertPendingApproval() *v1alpha1.UserSignup {
```
Comment (golangcibot) [inline test/e2e/user_management_test.go]: func `(*userManagementIntegrationTest).createUserSignupAndAssertPendingApproval` is unused (from `unused`)

File: test/e2e/user_management_test.go [hunk 12]
```diff
@@ -0,0 +1,122 @@
...
+		require.NoError(s.T(), err)
+	})
+}
+
+func (s *userManagementIntegrationTest) createUserSignupAndAssertAutoApproval(specApproved bool) (*v1alpha1.UserSignup, *v1alpha1.MasterUserRecord) {
+	id := uuid.NewV4().String()
+	return s.createAndCheckUserSignup(specApproved, "testuser"+id, "testuser"+id+"@test.com", approvedAutomatically()...)
+}
+
+func newBannedUser(host *wait.HostAwaitility, email string) *v1alpha1.BannedUser {
```
Comment (golangcibot) [inline test/e2e/user_management_test.go]: `newBannedUser` - `host` is unused (from `unparam`)

File: test/e2e/base_user_test.go [hunk 13]
```diff
@@ -0,0 +1,125 @@
...
+	userSignup := s.createAndCheckUserSignupNoMUR(specApproved, username, email, conditions...)
+
+	// Confirm the MUR was created
+	mur := s.assertCreatedMUR(userSignup)
+
+	return userSignup, mur
+}
+
+func (s *baseUserIntegrationTest) createAndCheckUserSignupNoMUR(specApproved bool, username string, email string,
+	conditions ...v1alpha1.Condition) (*v1alpha1.UserSignup) {
```
Comment (golangcibot) [inline test/e2e/base_user_test.go]: File is not `gofmt`-ed with `-s` (from `gofmt`)
```suggestion
	conditions ...v1alpha1.Condition) *v1alpha1.UserSignup {
```

File: test/e2e/user_management_test.go [hunk 14]
```diff
@@ -0,0 +1,112 @@
...
+	md5hash := md5.New()
+	_, _ = md5hash.Write([]byte(email))
+	emailHash := hex.EncodeToString(md5hash.Sum(nil))
+
+	return &v1alpha1.BannedUser{
+		ObjectMeta: v1.ObjectMeta{
+			Name: uuid.NewV4().String(),
```
Comment (golangcibot) [inline test/e2e/user_management_test.go]: File is not `gofmt`-ed with `-s` (from `gofmt`)
```suggestion
			Name:      uuid.NewV4().String(),
```

File: test/e2e/user_management_test.go [hunk 15]
```diff
@@ -0,0 +1,148 @@
+package e2e
+
+import (
+	"context"
+	"crypto/md5"
+	"encoding/hex"
+	authsupport "github.com/codeready-toolchain/toolchain-common/pkg/test/auth"
```
Comment (golangcibot) [inline test/e2e/user_management_test.go]: File is not `gofmt`-ed with `-s` (from `gofmt`)
```suggestion

```

File: test/e2e/user_management_test.go [hunk 16]
```diff
@@ -42,6 +46,96 @@ func (s *userManagementTestSuite) TearDownTest() {
...
+		s.setApprovalPolicyConfig("automatic")
+
+		// Create a new UserSignup and confirm it was approved automatically
+		userSignup, _ := s.createUserSignupAndAssertAutoApproval(false)
+
+		// Create the BannedUser
+		s.createAndCheckBannedUser(userSignup.Annotations[v1alpha1.UserSignupUserEmailAnnotationKey])
+
+		// Confirm the user is banned
+		s.hostAwait.WithRetryOptions(wait.TimeoutOption(time.Second * 10)).WaitForUserSignup(userSignup.Name,
```
Comment (golangcibot) [inline test/e2e/user_management_test.go]: Error return value of `(*github.com/codeready-toolchain/toolchain-e2e/wait.HostAwaitility).WaitForUserSignup` is not checked (from `errcheck`)

File: test/e2e/user_management_test.go [hunk 17]
```diff
@@ -9,24 +9,28 @@ import (
...
 	userv1 "github.com/openshift/api/user/v1"
-	framework "github.com/operator-framework/operator-sdk/pkg/test"
 	"github.com/stretchr/testify/assert"
 	"github.com/stretchr/testify/require"
 	"github.com/stretchr/testify/suite"
 	apierros "k8s.io/apimachinery/pkg/api/errors"
 	"k8s.io/apimachinery/pkg/types"
+	"crypto/md5"
+	"encoding/hex"
+	authsupport "github.com/codeready-toolchain/toolchain-common/pkg/test/auth"
```
Comment (golangcibot) [inline test/e2e/user_management_test.go]: File is not `gofmt`-ed with `-s` (from `gofmt`)
```suggestion
	authsupport "github.com/codeready-toolchain/toolchain-common/pkg/test/auth"
	userv1 "github.com/openshift/api/user/v1"
```

File: test/e2e/user_management_test.go [hunk 18]
```diff
@@ -42,6 +46,96 @@ func (s *userManagementTestSuite) TearDownTest() {
...
+		// Create the BannedUser
+		s.createAndCheckBannedUser(userSignup.Annotations[v1alpha1.UserSignupUserEmailAnnotationKey])
+
+		// Confirm the user is banned
+		s.hostAwait.WithRetryOptions(wait.TimeoutOption(time.Second * 10)).WaitForUserSignup(userSignup.Name,
+			wait.UntilUserSignupHasConditions(approvedAutomaticallyAndBanned()...))
+
+		// Confirm that a MasterUserRecord is deleted
+		mur, err := s.hostAwait.WithRetryOptions(wait.TimeoutOption(time.Second * 10)).WaitForMasterUserRecord(userSignup.Spec.Username)
+		log.Infof("### Found MUR: ", mur.Name)
```
Comment (golangcibot) [inline test/e2e/user_management_test.go]: printf: Infof call has arguments but no formatting directives (from `govet`)

File: test/e2e/user_management_test.go [hunk 19]
```diff
@@ -42,6 +46,96 @@ func (s *userManagementTestSuite) TearDownTest() {
...
+		s.setApprovalPolicyConfig("automatic")
+
+		// Create a new UserSignup and confirm it was approved automatically
+		userSignup, _ := s.createUserSignupAndAssertAutoApproval(false)
+
+		// Create the BannedUser
+		s.createAndCheckBannedUser(userSignup.Annotations[v1alpha1.UserSignupUserEmailAnnotationKey])
+
+		// Confirm the user is banned
+		s.hostAwait.WithRetryOptions(wait.TimeoutOption(time.Second*10)).WaitForUserSignup(userSignup.Name,
```
Comment (golangcibot) [inline test/e2e/user_management_test.go]: Error return value of `(*github.com/codeready-toolchain/toolchain-e2e/wait.HostAwaitility).WaitForUserSignup` is not checked (from `errcheck`)
